torch/fx/passes/infra/pass_manager.py

```diff
@@ -0,0 +1,169 @@
+    ):
+        if passes:
+            self.passes = passes
+        if constraints:
+            self.constraints = constraints
+        if steps:
+            self.steps = steps
+
+        self.run_checks_after_each_pass = run_checks_after_each_pass,
+
+    def add_pass(self, _pass: Callable):
+        """
+        Adds a pass into the current list of passes.
+        """
+        self.passes.append(_pass)
+        self._validated = False
+
+    def add_constraint(self, constraint: Callable):
+        """
+        Adds a constraint into the current list of constraints.
+        """
+        self.constraints.append(constraint)
+        self._validated = False
+
+    def validate_constraints(self):
+        """
+        Validates that current pass schedule defined by `self.passes` is valid
+        according to all constraints in `self.constraints`
+        """
+        if self._validated:
+            return
+        for constraint in self.constraints:
+            _validate_pass_schedule_constraint(constraint, self.passes)
+        self._validated = True
+
+    def add_checks(self, check: Callable) -> None:
+        """
+        Adds a function which takes runs various checks on a given graph module.
+        This function is run before and after each pass if the
+        `run_checks_after_each_pass` flag is enabled.
+        """
+        sig = inspect.signature(check)
+
+        if len(list(sig.parameters.values())) != 1:
+            raise TypeError("PassManager check function should only take in one variable, a graph_module")
+
+        setattr(self, "check", check)  # noqa: B010
+
+    def check(self, graph_module: GraphModule) -> None:
+        pass
+
+    def __call__(self, graph_module: GraphModule) -> None:
+        """
+        Runs a list of passes in the order based on `self.passes` on the given
+        graph module. Each time a pass is run, checks and linting will be run on
+        the graph module to ensure that it still maintains the same required
+        invariants.
+
+        The list of passes will be run until the graph stops changing, or until
+        `steps` number of times.
+        """
+        # Check constraints
+        self.validate_constraints()
+
+        # Lint and check graph invariants
+        graph_module.graph.lint()
+        self.check(graph_module)
+
+        # Run the set of passes `steps` number of times or until the graph stops
+        # changing
+        for _ in range(self.steps):
+            orig_graph_module_code = graph_module.code
+
+            # Run the set of passes on the graph module
+            for fn in self.passes:
+                fn(graph_module)
+
+                if self.run_checks_after_each_pass:
+                    # Lint and check graph invariants
+                    graph_module.graph.lint()
+                    self.check(graph_module)
+
+            graph_module.recompile()
+
+            # If the graph no longer changes, then we can stop running these passes
```

Comment:
I wonder if this is a reliable way to check if graph has been modified.... 
I think we PassBase should return a boolean, say "modified", to indicate if it actually took effect on modifying the graph. 
This "modified" return field could also be used to collect Pass Statistics.

Reply:
I noticed that ONNX does this too (have the pass base return a `modified` boolean), but I wasn't sure if it would be too much work for users to have to manage this information when implementing their pass.

However, having the user track this state does seem like the most accurate way to check if the graph has been modified. Right now I'm checking if the `graph_module.code` values are the same -- justification being that `graph_module.recompile()` recompiles the `code` based on the `graph`, so I assume that if the graph changes then the code will also change, and if the graph doesn't change then the code will also not change. I'm not positive this is a full-proof solution though.

What do you think?

Reply:
I saw that in Torchscript passes, and it seems to me that users usually wouldn't have difficulty returning this changing status to framework. I think in the long term it might be more flexible if users can just return a PassResult type and `changed` can be one of the data fields.